src/cluster/connectionBuilder.js

```diff
@@ -22,27 +21,26 @@ module.exports = ({
   logger,
   instrumentationEmitter = null,
 }) => {
-  validateBrokers(brokers)
-
-  const shuffledBrokers = shuffle(brokers)
-  const size = brokers.length
   let index = 0
 
   return {
-    build: ({ host, port, rack } = {}) => {
+    build: async ({ host, port, rack } = {}) => {
       if (!host) {
-        // Always rotate the seed broker
-        const [seedHost, seedPort] = shuffledBrokers[index++ % size].split(':')
-        host = seedHost
-        port = Number(seedPort)
+        const list = typeof brokers === 'function' ? await brokers() : brokers
+        validateBrokers(list)
+
+        const randomBroker = list[index++ % list.length]
+
+        host = randomBroker.split(':')[0]
+        port = Number(randomBroker.split(':')[1])
       }
 
       return new Connection({
         host,
         port,
         rack,
+        sasl: typeof sasl === 'function' ? await sasl() : sasl,
```

Comment:
Could we remove the sasl renewal functionality from this PR and handle it separately instead. I don't feel like I've fully thought through the ramifications of this, and once we make it a public interface, it's difficult to change if we later realize it should change.

For example, I'm not sure that on connection build is the right time to do this. If this call fails, how should we handle that? It may very well be that this is the way we should handle it, but I'd rather not have to merge both changes in the same PR.

Reply:
yes, removed